examples/src/bin/composite_template.rs

```diff
@@ -0,0 +1,106 @@
+//! # Composite Template Example
+//!
+//! This sample demonstrates how to create a widget using GTK's composite templates.
+
+#[macro_use]
+extern crate glib;
+#[macro_use]
+extern crate gtk;
+
+use gio::{prelude::ApplicationExtManual, ApplicationExt};
+use glib::subclass;
+use glib::subclass::prelude::*;
+use glib::{Cast, StaticType};
+use gtk::subclass::prelude::*;
+use gtk::subclass::widget::*;
+use gtk::WidgetExt;
+
+/// The private struct, which can hold widgets and other data.
+#[derive(Debug, CompositeTemplate)]
+pub struct ExApplicationWindowPriv {
+    // The #[template_child] attribute tells the CompositeTemplate macro
+    // that a field is meant to be a child within the template.
+    #[template_child(id = "headerbar")]
+    headerbar: TemplateChild<gtk::HeaderBar>,
+    #[template_child(id = "label")]
+    label: TemplateChild<gtk::Label>,
+    #[template_child(id = "subtitle_label")]
+    subtitle: TemplateChild<gtk::Label>,
+}
+
+impl ObjectSubclass for ExApplicationWindowPriv {
+    const NAME: &'static str = "ExApplicationWindow";
+    type Type = ExApplicationWindow;
+    type ParentType = gtk::ApplicationWindow;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self {
+            headerbar: TemplateChild::default(),
+            label: TemplateChild::default(),
+            subtitle: TemplateChild::default(),
+        }
+    }
+
+    // Within class_init() you must set the template
+    // and bind it's children. The CompositeTemplate
+    // derive macro provides a convenience function
+    // bind_template_children() to bind all children
+    // at once.
+    fn class_init(klass: &mut Self::Class) {
+        let template = include_bytes!("composite_template.ui");
+        Self::set_template(klass, template);
+        Self::bind_template_children(klass);
+    }
+}
+
+impl ObjectImpl for ExApplicationWindowPriv {}
+impl WidgetImpl for ExApplicationWindowPriv {}
+impl WindowImpl for ExApplicationWindowPriv {}
+impl ApplicationWindowImpl for ExApplicationWindowPriv {}
+
+glib_wrapper! {
+    pub struct ExApplicationWindow(ObjectSubclass<ExApplicationWindowPriv>)
+        @extends gtk::Widget, gtk::Window, gtk::ApplicationWindow, @implements gio::ActionMap, gio::ActionGroup;
+}
+
+impl ExApplicationWindow {
+    pub fn new<P: glib::IsA<gtk::Application> + glib::value::ToValue>(app: &P) -> Self {
+        let win = glib::Object::new(Self::static_type(), &[("application", app)])
+            .expect("Failed to create ExApplicationWindow")
+            .downcast::<ExApplicationWindow>()
+            .expect("Created object is of wrong type");
+        // init_template() must be run after the object is constructed.
+        win.init_template();
+        win.init_label();
```

Comment:
Oh really? You can't do that in `ObjectImpl::constructed()` after `self.parent_constructed()`?

Reply:
Actually this is too late here:

> It is important to call this function in the instance initializer of a GtkWidget subclass and not in GObject.constructed() or GObject.constructor() for two reasons.

That means you must call it in `instance_init()`. There is currently no hook for this (and that would also be unsafe). See `glib/src/subclass/types.rs:instance_init()`.

Would it be a problem to always call that for widget subclasses automatically?

Reply:
I guess the best for now would be to call it from `ObjectImpl::constructed()` before `parent_constructed()`. That's as early as we can do currently.

And then create an issue for figuring out a way to call it even earlier.

Reply:
That means that you can also do `init_label()` from `constructed()` already btw

Reply:
That's annoying :)